shopfloor_reception/services/reception.py

```diff
@@ -222,54 +243,51 @@ def _select_line(self, picking, line, move, increase_qty_done_by=1):
+        product = line.product_id
         self._assign_user_to_picking(picking)
         self._assign_user_to_line(line)
-        line.qty_done += increase_qty_done_by
+        line.qty_done += qty_done
         if product.tracking not in ("lot", "serial") or (line.lot_id or line.lot_name):
             return self._response_for_set_quantity(picking, line)
         return self._response_for_set_lot(picking, line)
 
-    def _select_line_from_product(self, picking, move, product):
-        line = fields.first(
-            picking.move_line_ids.filtered(
-                lambda l: l.product_id == product
+    def _select_line__filter_lines_by_packaging__return(self, lines, packaging):
+        return_line = fields.first(
+            lines.filtered(
+                lambda l: not l.package_id.product_packaging_id
                 and not l.result_package_id
-                and l.shopfloor_user_id.id in [False, self.env.uid]
+                and l.shopfloor_user_id.id in (False, self.env.uid)
             )
         )
-        return self._select_line(picking, line, move)
+        if return_line:
+            return return_line
 
-    def _select_line_from_packaging(self, picking, move, packaging):
-        line = fields.first(
-            picking.move_line_ids.filtered(
-                lambda l: l.package_id.product_packaging_id == packaging
-                and not l.result_package_id
-                and l.shopfloor_user_id.id in [False, self.env.uid]
+    def _select_line__filter_lines_by_packaging(self, lines, packaging):
+        if self.work.menu.allow_return:
+            line = self._select_line__filter_lines_by_packaging__return(
+                lines, packaging
             )
-        )
-        return self._select_line(picking, line, move, packaging.qty)
-
-    def _select_line_from_lot(self, picking, move, lot):
-        line = fields.first(
-            picking.move_line_ids.filtered(
-                lambda l: (l.lot_id.name == lot or l.lot_name == lot)
+            if line:
+                return line
+        return fields.first(
+            lines.filtered(
+                lambda l: l.package_id.product_packaging_id == packaging
                 and not l.result_package_id
                 and l.shopfloor_user_id.id in [False, self.env.uid]
             )
         )
-        if not line:
-            return
-        return self._select_line(picking, line, move)
 
     def _order_stock_picking(self):
         # We sort by scheduled date first. However, there might be a case
         # where two pickings have the exact same scheduled date.
         # In that case, we sort by id.
         return "scheduled_date ASC, id ASC"
 
-    def _scan_document__by_picking(self, barcode):
-        search = self._actions_for("search")
-        picking_filter_result = search.picking_from_scan(barcode, use_origin=True)
+    def _scan_document__by_picking(self, pickings, barcode):
```

Comment:
Seems that `barcode` is not used anymore in this method now. Is it expected as the method is named `_scan_document__[...]`?

Reply:
yes, scan_document is the endpoint.
We already found a picking from the barcode.
To me it makes sense to name those methods `_<endpoint>__by_<model>`.
If we don't prefix by endpoint, we will end with a real mess, if for instance, a product can be scanned on every screen.

Reply:
no, this is a handler method, therefore, it is named `_<endpoint>__by_<model>`, but the scan is already done and we've found a record already.

Reply:
But the barcode argument can be removed from the function declaration, no ?
Same question for the other following functions